File: src/main/java/com/ort/app/web/form/validator/RandomKeywordFormValidator.java
package com.ort.app.web.form.validator;

import java.util.stream.Stream;

import org.springframework.stereotype.Component;
import org.springframework.validation.Errors;
import org.springframework.validation.Validator;

import com.ort.app.web.form.RandomKeywordForm;

@Component
public class RandomKeywordFormValidator implements Validator {

    @Override
    public boolean supports(Class<?> paramClass) {
        return RandomKeywordForm.class.isAssignableFrom(paramClass);
    }

    @Override
    public void validate(Object paramObject, Errors errors) {
        if (errors.hasErrors()) {
            // 既にエラーがある場合はチェックしない
            return;
        }

        RandomKeywordForm form = (RandomKeywordForm) paramObject;

        String keyword = form.getKeyword();
        if (keyword != null && keyword.contains("or") || keyword.contains("who")) {
            errors.rejectValue("keyword", "RandomKeywordForm.validator.keyword.ngword");
        }

        String message = form.getMessage();
        if (message == null) {
            return;
        }
        String trimedMessage = message.trim();
        String[] messages = trimedMessage.split("\r\n");
        Stream.of(messages).forEach(mes -> {
            if (mes.length() < 1 || mes.length() > 20) {
                errors.rejectValue("message", "RandomKeywordForm.validator.message.length");
                return;
            }
        });
        for (int i = 0; i < messages.length - 1; i++) {
            for (int j = i + 1; j < messages.length; j++) {
                if (messages[i].equals(messages[j])) {
                    errors.rejectValue("message", "RandomKeywordForm.validator.message.duplicate");
                    return;
                }
            }
        }

    }
}

File: backend/src/main/kotlin/com/ort/app/api/request/validator/RandomKeywordFormValidator.kt
package com.ort.app.api.request.validator

import com.ort.app.api.request.RandomKeywordForm
import org.springframework.stereotype.Component
import org.springframework.validation.Errors
import org.springframework.validation.Validator

@Component
class RandomKeywordFormValidator : Validator {

    override fun supports(clazz: Class<*>): Boolean {
        return RandomKeywordForm::class.java.isAssignableFrom(clazz)
    }

    override fun validate(target: Any, errors: Errors) {
        if (errors.hasErrors()) return

        val form = target as RandomKeywordForm

        val keyword: String = form.keyword!!
        if (keyword.contains("or") || keyword.contains("who")) {
            errors.rejectValue("keyword", "RandomKeywordForm.validator.keyword.ngword")
        }

        val messages = form.message!!.trim().split("\r\n")
        if (messages.any { it.isEmpty() || it.length > 20 }) {
            errors.rejectValue("message", "RandomKeywordForm.validator.message.length")
        }
        if (messages.distinct().size != messages.size) {
            errors.rejectValue("message", "RandomKeywordForm.validator.message.duplicate")
        }
    }
}